### ISSUE TITLE
[HOLD for payment 2023-05-03] [$1000] Leading whitespaces aren’t truncated while saving email/phone number in  new contact method page

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17770]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to Settings -> Profile -> contact methods
2. Click on new contact methods
3. Add email with spaces at front e.g “                [sharmashim123@gmail.com](mailto:sharmashim123@gmail.com)”
4. Type in password and save
5. Now, it should redirect to contact methods page
6. Notice that the leading spaces aren’t removed while displaying email/phone number

## Expected Result:  
Leading or trailing spaces should be removed and email.phone number should be displayed accordingly for better UI
## Actual Result : 
Leading spaces makes improper indentation while display.
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/233529701-d5974fa3-e27f-4ce8-b726-f489357e43a3.mov

<img width="960" alt="Untitled" src="https://user-images.githubusercontent.com/43996225/233529705-644a61fb-6c6b-4226-842e-31a02b9b1a79.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ashimsharma10
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681998814607449

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b41c774d839dccd0</li>
        <li>Upwork Job ID: 1649571325356744704</li>
        <li>Last Price Increase: 2023-04-22</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @strepanier03 (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by PrashantMangukiya** [comment]
## Proposal

Posting proposal early as per new [guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job)

### Please re-state the problem that we are trying to solve in this issue.
Leading whitespaces aren not truncated while saving email/phone number in new contact method page.

### What is the root cause of that problem?
We can see within `onLoginChange()` we do not trim `login` before save to state, so login value will be with leading and trailing space if entered.
https://github.com/Expensify/App/blob/4f7cfdb53216185ac7e825127b3be2f8238340fd/src/pages/settings/Profile/Contacts/NewContactMethodPage.js#L68-L70

Also during `submitForm()` we do not trim `this.state.login` as shown below,
https://github.com/Expensify/App/blob/4f7cfdb53216185ac7e825127b3be2f8238340fd/src/pages/settings/Profile/Contacts/NewContactMethodPage.js#L85-L92

So `User.addNewContactMethodAndNavigate()` method will save `login` value with leading and trailing space. This is the root cause of the problem.

### What changes do you think we should make in order to solve the problem?

To solve this we can think to trim login while save to state at line 69 within `onLoginChange()`. 

**But it is not advisable** because user may enter **Email** Or **Phone number** within login field. So if user typing email then trim() during save to state will be ok, because generally blank space not comes within email so user will not try to enter blank space within email.

**But if user enter Phone number then user have habit to enter phone number like "123 4567 8901" i.e. with space.** So if we trim value during onChange, then as soon as user type space within phone number onLoginChange() will trim space during save to state. So it will make bad user experience and user will not able to enter space within phone number.

**So better approach is to trim login within `submitForm()` function** and use trimmed login value for `lodashGet()` and `addNewContactMethodAndNavigate()` api. So modified `submitForm()` function code looks as shown below.
```javascript
submitForm() {
    // Trim leading and trailing space from login
    const login = this.state.login.trim();

    // If this login already exists, just go back.
    if (lodashGet(this.props.loginList, login)) {
        Navigation.navigate(ROUTES.SETTINGS_CONTACT_METHODS);
        return;
    }
    User.addNewContactMethodAndNavigate(login, this.state.password);
}
```

So this will solve the issue i.e. truncate leading/trailing white space from email or phone number, and allows to enter space within phone number. Result video attached below.

### What alternative solutions did you explore? (Optional)
None.

<details>
<summary>Results</summary>

#### Truncate leading/trailing space from email.

https://user-images.githubusercontent.com/7823358/233578309-69a620dd-a2c2-4ce9-9abf-2213d26af09d.mov

#### Truncate leading/trailing space from Phone and allow space in between phone number.

https://user-images.githubusercontent.com/7823358/233578399-ebfc37b9-db7c-47b7-ad1c-073f595b47a1.mov

https://user-images.githubusercontent.com/7823358/233585221-7127fade-4899-4976-86a8-5a1c5dd86f2a.mov

</details>

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by strepanier03** [comment]
Able to recreate following instructions but oddly I have the email listed twice, once with the extra spaced and once without.

<img width="1081" alt="2023-04-21_17-27-15" src="https://user-images.githubusercontent.com/10925636/233752037-0dce2d6d-0b3a-4ca6-9ecf-4f2f6accd8b5.png">

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01b41c774d839dccd0

**Comment by MelvinBot** [comment]
Current assignee @strepanier03 is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @Santhosh-Sellavel (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @luacmartins (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Santhosh-Sellavel** [comment]
@PrashantMangukiya's [Proposal](https://github.com/Expensify/App/issues/17770#issuecomment-1517469634) LGTM, straight forward change!

**Comment by MelvinBot** [comment]
📣 @PrashantMangukiya You have been assigned to this job by @luacmartins!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by Santhosh-Sellavel** [comment]
> @Santhosh-Sellavel @luacmartins If we allow space on the phone number then case I think the issue will remain the same for the phone right?

It won't create the current issue, we are just trying to trim out leading and trailing space. So we are good here.

If you are talking about space between the phone numbers, that's an invalid phone number case we got already covered with showing an error. So we are good to go here.

But restricting user input is bad UX.

**Comment by PrashantMangukiya** [comment]
@Santhosh-Sellavel  @luacmartins  Thank you. Applied to job on Upwork. I will prepare and submit PR within 3 hours asap.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.5-6 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17817


If no regressions arise, payment will be issued on 2023-05-03. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@Santhosh-Sellavel] Determine if we should create a regression test for this bug.
- [ ] [@Santhosh-Sellavel] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@strepanier03] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by strepanier03** [comment]
Just getting back from out of office and working to catch up on this now. I'll have an update shortly.

**Comment by strepanier03** [comment]
@ashimsharma10 - I have hired you to a bug report job in Upwork.

@Santhosh-Sellavel and @PrashantMangukiya - I have hired you to the Upwork job as well. 

----

@Santhosh-Sellavel - If you can finish the BZ checklist I'll update the comment and make any necessary reg test GHs.

----

I'll check in later today to pay what I can.

**Comment by PrashantMangukiya** [comment]
@strepanier03 Offer accepted on Upwork.

**Comment by strepanier03** [comment]
@PrashantMangukiya and @ashimsharma10, I have paid you both via Upwork and closed contracts, thank you both!

**Comment by PrashantMangukiya** [comment]
@strepanier03  Received, Thank you.

**Comment by Santhosh-Sellavel** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [x]  [@Santhosh-Sellavel] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/16919
> * [ ]  [@Santhosh-Sellavel] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
> * [ ]  [@Santhosh-Sellavel] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
> * [ ]  [@Santhosh-Sellavel] Determine if we should create a regression test for this bug.
> * [ ]  [@Santhosh-Sellavel] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
> * [ ]  [@strepanier03] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by Santhosh-Sellavel** [comment]
### Regression steps

From the PR itself looks good to me

1. Go to Settings -> Profile -> Contact methods
2. Click on "New contact method" button
3. Add email/phone with leading and trailing spaces e.g “ [test@test.com](mailto:test@test.com) ”
4. Tap Add button
5. Now, it will go back to "Contact methods" page
6. Verify that leading/trailing spaces does not show while display email/phone number.

👍  or  👎 

cc: @luacmartins or @strepanier03

**Comment by Santhosh-Sellavel** [comment]
@strepanier03 Accepted offer thanks!
